### PR TITLE
Move unit tests to separate directories, and document

### PR DESCRIPTION
**Pull request by Ericson2314**
# Motivation

Today, with the tests inside a `tests` intermingled with the corresponding library's source code, we have a few problems:

- We have to be careful that wildcards don't end up with tests being built as part of Nix proper, or test headers being installed as part of Nix proper.

- Tests in libraries but not executables is not right:

  - It means each executable runs the previous unit tests again, because
    it needs the libraries.

  - It doesn't work right on Windows, which doesn't want you to load a DLL just for the side global variable . It could be made to work with the dlopen equivalent, but that's gross!

This reorg solves these problems.

There is a remaining problem which is that sibbling headers (like `hash.hh` the test header vs `hash.hh` the main `libnixutil` header) end up shadowing each other. This PR doesn't solve that. That is left as future work for a future PR.

# Context

This does something more minimal than what #7876 asks for, but it is a step in the right direction.

# Checklist for maintainers

Maintainers: tick if completed or explain if not relevant

 - [ ] agreed on idea
 - [ ] agreed on implementation strategy
 - [ ] tests, as appropriate
   - functional tests - `tests/**.sh`
   - unit tests - `src/*/tests`
   - integration tests - `tests/nixos/*`
 - [ ] documentation in the manual
 - [ ] documentation in the internal API docs
 - [ ] code and comments are self-explanatory
 - [ ] commit message explains why the change was made
 - [ ] new feature or incompatible change: updated release notes

# Priorities

Add :+1: to [pull requests you find important](https://github.com/NixOS/nix/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc).

### CODE REVIEW

**Comment by Ericson2314** [comment]
I should add that the way things are today was quite concerning to @p01arst0rm when working on the Meson fork.

**Comment by p01arst0rm** [comment]
good PR 👍

File: doc/manual/src/contributing/testing.md [hunk 1]
```diff
@@ -2,10 +2,56 @@
...
+> │   │   │
+> │   │   …
+> │   ├── value/context.cc
+> │   │
+> │   …
+> …
+> ```
+
+The tests for each Nix library (`libnixexpr`, `libnixstore`, etc..) live inside a directory `src/${library_name_without-nix}-tests` next to the directory for the library (`src/${library_name_without-nix}`).
+Given a `foo/bar.{hh,cc}` interface (header) and implementation pair in the original library, we write tests for it in `foo/bar.cc`, and (possibly) declare additional interfaces in `tests/foo/bar.cc` in the test directory.
```
Comment (cole-h) [inline doc/manual/src/contributing/testing.md:33]: Rather than the arbitrary `foo` and `bar` here, could you maybe reword this so it picks up the naming in the tree above? That would make it much more concrete, and much easier for me to understand.

Specifically, why is there a `libexpr-tests/value/context.cc` (I imagine this is just the tests for `libexpr/value/context.cc`) as well as a `libexpr-tests/tests/value/context.hh` (what's the difference between this and `libexpr/value/context.hh`)?

Comment (p01arst0rm) [reply doc/manual/src/contributing/testing.md:33]: the fix here is to prevent name collisions in the first place by separating out the directories, its less a case of concerning the contents of the files.

Comment (Ericson2314) [reply doc/manual/src/contributing/testing.md:33]:
```suggestion
Given a interface (header) and implementation pair in the original library, say, `value/context.{hh,cc}`, we write tests for it in `value/context.cc`, and (possibly) declare additional interfaces for testing purposes in `test/value/context.hh`.
(Note only the extra header is nested inside the `test` directory. This is explained further below.) 
```
How's this?

Comment (cole-h) [reply doc/manual/src/contributing/testing.md:33]: Yes, that is much easier for me to wrap my head around! Thanks!

File: doc/manual/src/contributing/testing.md [hunk 2]
```diff
@@ -2,10 +2,57 @@
 
 ## Unit-tests
 
-The unit-tests for each Nix library (`libexpr`, `libstore`, etc..) are defined
-under `src/{library_name}/tests` using the
-[googletest](https://google.github.io/googletest/) and
-[rapidcheck](https://github.com/emil-e/rapidcheck) frameworks.
+The unit-tests for defined using the [googletest] and [rapidcheck] frameworks.
```
Comment (fricklerhandwerk) [inline doc/manual/src/contributing/testing.md:5]:
```suggestion
The unit-tests are defined using the [googletest] and [rapidcheck] frameworks.
```

File: doc/manual/src/contributing/testing.md [hunk 3]
```diff
@@ -2,10 +2,57 @@
...
+Given a interface (header) and implementation pair in the original library, say, `value/context.{hh,cc}`, we write tests for it in `value/context.cc`, and (possibly) declare additional interfaces for testing purposes in `test/value/context.hh`.
+(Note only the extra header is nested inside the `test` directory. This is explained further below.)
+
+The use of a separate directory for the unit tests might seem inconvenient, as the tests are not "right next to" the part of the code they are testing.
+But organizing the tests this way has one big benefit:
+there is no risk of any build-system wildcards for the library accidentally picking up test code that shoud not built and installed as part of the library.
+
+Likewise, the use of the `test/` subdir might seem superfluous:
+Isn't the point of the `*-test` subdir to indicate that these files are tests?
+Why do we need another `test` subdir too?
```
Comment (fricklerhandwerk) [inline doc/manual/src/contributing/testing.md:42]:
```suggestion
Why do we need another `test` subdirectory?
```

File: doc/manual/src/contributing/testing.md [hunk 4]
```diff
@@ -2,10 +2,57 @@
...
+The answer is that we need to be able to tell apart the two headers, and make sure we include the right one.
+For `.cc` files, this is matter of doing `#include "foo.hh"` vs `#include "tests/foo.hh`
+For `.hh` files, this is a bit more subtle, because a `#include "foo.hh` instead `tests/foo.hh` will end up including itself because `#include "..."`
+[always prioritizes files in the same directory](https://pubs.opengroup.org/onlinepubs/9699919799/utilities/c99.html#tag_20_11_04), rather than the original header from the library.
+Instead, use `#include <foo.hh>` to get the original library header, and `#include "tests/foo.hh"` to get the test header.
+
+Why do we have test headers at all?
+These are usually for [rapidcheck]'s `Arbitrary` machinery, which is used to describe how to generate values of a given type for the sake of running property tests.
+Because types contain other types, `Arbitrary "instances" for some type are not just useful for testing that type, but also any other type that contains it.
```
Comment (fricklerhandwerk) [inline doc/manual/src/contributing/testing.md:51]:
```suggestion
Because types contain other types, `Arbitrary` "instances" for some type are not just useful for testing that type, but also any other type that contains it.
```

File: doc/manual/src/contributing/testing.md [hunk 5]
```diff
@@ -2,10 +2,57 @@
...
+> │   │   ├── value/context.hh
+> │   │   │
+> │   │   …
+> │   ├── value/context.cc
+> │   │
+> │   …
+> …
+> ```
+
+The tests for each Nix library (`libnixexpr`, `libnixstore`, etc..) live inside a directory `src/${library_name_without-nix}-tests` next to the directory for the library (`src/${library_name_without-nix}`).
```
Comment (fricklerhandwerk) [inline doc/manual/src/contributing/testing.md:32]: Scrolling back and forth I understand what you mean by "without-nix", but would it hurt so much to not have that distinction? E.g. by removing the nix in the original library name or adding it for the tests as well? It makes something that's fairly involved and obscure needlessly hard to grok.

File: doc/manual/src/contributing/testing.md [hunk 6]
```diff
@@ -2,10 +2,57 @@
...
+Given a interface (header) and implementation pair in the original library, say, `value/context.{hh,cc}`, we write tests for it in `value/context.cc`, and (possibly) declare additional interfaces for testing purposes in `test/value/context.hh`.
```
Comment (fricklerhandwerk) [inline doc/manual/src/contributing/testing.md:33]:
```suggestion
Given a interface (header) and implementation pair in the original library, say, `libexpr/value/context.{hh,cc}`, we write tests for it in `libexpr-tests/value/context.cc`, and (possibly) declare additional interfaces for testing purposes in `libexpr-tests/tests/value/context.hh`.
```

Is it test or tests now? I'd prefer test because it's fewer characters and doesn't lose relevant information.

File: doc/manual/src/contributing/testing.md [hunk 7]
```diff
@@ -2,10 +2,57 @@
...
+(Note only the extra header is nested inside the `test` directory. This is explained further below.)
```
Comment (fricklerhandwerk) [inline doc/manual/src/contributing/testing.md:34]:
```suggestion

> **Note**
>
> Only the extra header is nested inside the `test` directory. This is explained further below.
```

If you have an anchor there you can link directly.

File: doc/manual/src/contributing/testing.md [hunk 8]
```diff
@@ -2,10 +2,57 @@
...
+
+The use of a separate directory for the unit tests might seem inconvenient, as the tests are not "right next to" the part of the code they are testing.
```
Comment (fricklerhandwerk) [inline doc/manual/src/contributing/testing.md:36]:
```suggestion
### Rationale

The use of a separate directory for the unit tests might seem inconvenient, as the tests are not "right next to" the part of the code they are testing.
```

Breaking this long text down with headings will make it easier to navigate. You can use the leading questions as level 4 headers.

File: doc/manual/src/contributing/testing.md [hunk 9]
```diff
@@ -2,10 +2,57 @@
...
+The answer is that we need to be able to tell apart the two headers, and make sure we include the right one.
+For `.cc` files, this is matter of doing `#include "foo.hh"` vs `#include "tests/foo.hh`
+For `.hh` files, this is a bit more subtle, because a `#include "foo.hh` instead `tests/foo.hh` will end up including itself because `#include "..."`
+[always prioritizes files in the same directory](https://pubs.opengroup.org/onlinepubs/9699919799/utilities/c99.html#tag_20_11_04), rather than the original header from the library.
+Instead, use `#include <foo.hh>` to get the original library header, and `#include "tests/foo.hh"` to get the test header.
+
+Why do we have test headers at all?
+These are usually for [rapidcheck]'s `Arbitrary` machinery, which is used to describe how to generate values of a given type for the sake of running property tests.
+Because types contain other types, `Arbitrary "instances" for some type are not just useful for testing that type, but also any other type that contains it.
+(Indeed, if we don't reuse the upstream type's `Arbitrary` instance, the downstream type's `Arbitrary` instance would become much more complex and hard to understand.)
```
Comment (fricklerhandwerk) [inline doc/manual/src/contributing/testing.md:52]:
```suggestion
Indeed, if we don't reuse the upstream type's `Arbitrary` instance, the downstream type's `Arbitrary` instance would become much more complex and hard to understand.
```

This is a proper argument, no need to parenthesise.

File: doc/manual/src/contributing/testing.md (hunk 6 above)
Comment (Ericson2314) [reply doc/manual/src/contributing/testing.md:33]: Oh oops, should be `test`

File: doc/manual/src/contributing/testing.md (hunk 5 above)
Comment (p01arst0rm) [reply doc/manual/src/contributing/testing.md:32]: that was actually discussed as a future goal of current refactoring efforts :) the problem currently is striking a balance between discrete specific small improvements that can be introduced overtime, and a just one huge commit that changes everything. seeing as the current layout uses libnix* to discuss nix libraries, it would be best to keep as is for now imo

Comment (Ericson2314) [reply doc/manual/src/contributing/testing.md:32]: Yeah the other dir is `libexpr` so it seems prudent to let that be for now and call this `libexpr-tests`

**Review by fricklerhandwerk — APPROVED** [review]

File: doc/internal-api/doxygen.cfg.in [hunk 10]
```diff
@@ -39,17 +39,21 @@ INPUT                  = \
   src/libcmd \
   src/libexpr \
   src/libexpr/flake \
-  src/libexpr/tests \
-  src/libexpr/tests/value \
+  src/libexpr-tests \
+  src/libexpr-tests/value \
+  src/libexpr-tests/test \
+  src/libexpr-tests/test/value \
```
Comment (fricklerhandwerk) [inline doc/internal-api/doxygen.cfg.in:45]: Discussed in the team meeting:
```suggestion
  tests/libexpr/tests/value \
```

While the nested `tests` looks silly, it makes the diff within the files smaller.

**Comment by fricklerhandwerk** [comment]
Discussed in Nix team meeting:

- moving the nested `tests` -> `test` produces somewhat unnecessary diff
- @thufschmitt: is there any cost to nesting all of that under a top-level `tests` directory?
  - @Ericson2314: can be done
- assigned to @thufschmitt

**Comment by Ericson2314** [comment]
Somewhat surprisingly, this has become relevant for the Windows work (I need to separate the test exes and tests libraries better, `tests/foo.cc` for lib vs `foo.cc` will do the trick).

Let me know if this looks good to you, @thufschmitt. (The conflicts are usually extremely superficial and easy to fix.)

File: doc/manual/src/contributing/testing.md [hunk 11]
```diff
@@ -30,32 +30,69 @@ The unit tests are defined using the [googletest] and [rapidcheck] frameworks.
...
+> │           │   …
+> │           ├── libexpr
+> │           │   │
+> │           │   …
+> │           …
 > …
 > ```
 
-The unit tests for each Nix library (`libnixexpr`, `libnixstore`, etc..) live inside a directory `src/${library_shortname}/tests` within the directory for the library (`src/${library_shortname}`).
+The tests for each Nix library (`libnixexpr`, `libnixstore`, etc..) live inside a directory `src/${library_name_without-nix}-tests` next to the directory for the library (`src/${library_name_without-nix}`).
```
Comment (thufschmitt) [inline doc/manual/src/contributing/testing.md:60]: This is now wrong, right?

File: doc/manual/src/contributing/testing.md [hunk 12]
```diff
@@ -30,32 +30,69 @@ The unit tests are defined using the [googletest] and [rapidcheck] frameworks.
...
+For `.cc` files, this is matter of doing `#include "foo.hh"` vs `#include "tests/foo.hh`
+For `.hh` files, this is a bit more subtle, because a `#include "foo.hh` instead `test/foo.hh` will end up including itself because `#include "..."`
+[always prioritizes files in the same directory](https://pubs.opengroup.org/onlinepubs/9699919799/utilities/c99.html#tag_20_11_04), rather than the original header from the library.
+Instead, use `#include <foo.hh>` to get the original library header, and `#include "tests/foo.hh"` to get the test header.
+
+Why do we have test headers at all?
+These are usually for [rapidcheck]'s `Arbitrary` machinery, which is used to describe how to generate values of a given type for the sake of running property tests.
+Because types contain other types, `Arbitrary` "instances" for some type are not just useful for testing that type, but also any other type that contains it.
+Indeed, if we don't reuse the upstream type's `Arbitrary` instance, the downstream type's `Arbitrary` instance would become much more complex and hard to understand.
+In order to reuse these instances, we therefore declare them in these testing headers.
```
Comment (thufschmitt) [inline doc/manual/src/contributing/testing.md:95]: Is don't think we care a bout that at all here. This is nice for a commit message, but users/contributors generally shouldn't be really concerned about it, and wouldn't read it any way.

**Review by thufschmitt — APPROVED** [review]
A couple of small things in the testing manual, but looks good otherwise.

Feel free to merge once it's addressed (or discarded, I don't feel too strongly about these)

File: doc/manual/src/contributing/testing.md (hunk 11 above)
Comment (Ericson2314) [reply doc/manual/src/contributing/testing.md:60]: Good catch. There are a few things that need to be fixed here.

File: doc/manual/src/contributing/testing.md (hunk 12 above)
Comment (fricklerhandwerk) [reply doc/manual/src/contributing/testing.md:95]: I think it's the generally he right place for this kind of contextual information, but we may consider moving that out of the manual and into the tree.

Comment (Ericson2314) [reply doc/manual/src/contributing/testing.md:95]: I want contributors to also write property tests and "Arbitrary instances", so I do want that pattern to be documented.

Comment (Ericson2314) [reply doc/manual/src/contributing/testing.md]: Amid many other changes this text is now reorganized out of the file layout section. Hope that is better.

**Comment by Ericson2314** [comment]
@thufschmitt I am embarrassed to say that my "I'll just fix conflicts after" was wrong, so I think this needs to be re-reviewed.

Here is what happened:

- I realized putting the source files also in `tests` defeated one of the purposes purpose of the original PR, which is to make sure that `#include "foo.h"` without `test/` unambiguously refereed to the original library header, and not the test header. *Nevertheless*, I kept the companions to `test` headers in `test` --- i.e. the PR does the thing we agreed on the meeting despite these issues and instead no longer attempts to address this ambiguous `#include` problem.

- I better separate the "test libs" with headers and arbitrary instances from the tests themselves without headers:
  - libraries get their own `test/unit/*-lib` dir.
  - Now most tests `.cc` files do not need to go in `tests/` because they don't correspond to a header
  - This fixes running the unit tests on windows, where it doesn't want to load a separate DLL just for the side effects.
- unit test data dirs are not longer a separate hierarchy, as avoiding "cumulative" test exes means we don't need a single root data dir.
  
So one problem is no longer solved in the name of concensus, but another problem is solved (the windows problem), so this PR is hopefully still just as net useful.
